d2/src/main/java/com/linkedin/d2/xds/XdsToD2PropertiesAdaptor.java

```diff
@@ -302,19 +276,138 @@ private ClusterProperties toClusterProperties(Struct clusterProperties, long ver
         JsonFormat.printer().print(clusterProperties).getBytes(StandardCharsets.UTF_8), version);
   }
 
-  private UriProperties toUriProperties(String clusterName, Map<String, XdsD2.D2Node> uriDataMap)
+  private Map<String, UriProperties> toUriProperties(Map<String, XdsD2.D2Node> uriDataMap)
       throws InvalidProtocolBufferException, PropertySerializationException
   {
-    List<UriProperties> allUriProperties = new ArrayList<>();
+    Map<String, UriProperties> parsedMap = new HashMap<>();
 
-    for (XdsD2.D2Node d2Node : uriDataMap.values())
+    for (Map.Entry<String, XdsD2.D2Node> entry : uriDataMap.entrySet())
     {
+      XdsD2.D2Node d2Node = entry.getValue();
       UriProperties uriProperties = _uriPropertiesJsonSerializer.fromBytes(
           JsonFormat.printer().print(d2Node.getData()).getBytes(StandardCharsets.UTF_8), d2Node.getStat().getMzxid());
-      allUriProperties.add(uriProperties);
+      parsedMap.put(entry.getKey(), uriProperties);
+    }
+
+    return parsedMap;
+  }
+
+  private class UriPropertiesResourceWatcher implements XdsClient.D2NodeMapResourceWatcher
+  {
+    final String _clusterName;
+    final AtomicBoolean _init;
+    final long _initFetchStart;
+
+    Map<String, UriProperties> _currentData = new HashMap<>();
+
+    public UriPropertiesResourceWatcher(String clusterName)
+    {
+      _clusterName = clusterName;
+      _init = new AtomicBoolean();
+      _initFetchStart = System.nanoTime();
+    }
+
+    @Override
+    public void onChanged(XdsClient.D2NodeMapUpdate update)
+    {
+      if (_init.compareAndSet(false, true))
+      {
+        emitSDStatusInitialRequestEvent(_clusterName);
+      }
+
+      if (_uriEventBus != null)
+      {
+        try
+        {
+          Map<String, UriProperties> updates = toUriProperties(update.getNodeDataMap());
+          emitSDStatusUpdateReceiptEvents(updates);
```

Comment:
as we discussed in the standup, if this is from initial request response, we don't want to emit update receipt events to avoid the false spike in propagation latency (because the servers could be announced very long time ago before the client comes up).